src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs

```diff
@@ -322,13 +322,35 @@ private static string GenerateUpgradeReport(NuGetProject nuGetProject, string ba
             bool acceptedLicense = true;
 
             List<string> removedPackages = null;
+            List<Tuple<string, string>> existingPackages = new List<Tuple<string, string>>();
             List<Tuple<string,string>> addedPackages = null;
             List<Tuple<string,string>> updatedPackagesOld = null;
             List<Tuple<string,string>> updatedPackagesNew = null;
 
             // Enable granular level telemetry events for nuget ui operation
             uiService.ProjectContext.OperationId = Guid.NewGuid();
 
+            try
+            {
+                // collect the install state of the existing packages
+                foreach (var project in uiService.Projects)
+                {
+                    var result = await project.GetInstalledPackagesAsync(token);
+                    foreach (var package in result)
+                    {
+                        Tuple<string, string> packageInfo = new Tuple<string, string>(package.PackageIdentity.Id, (package.PackageIdentity.Version == null ? "" : package.PackageIdentity.Version.ToNormalizedString()));
+                        if (!existingPackages.Contains(packageInfo))
```

Comment:
I'm concerned how long `.Contains` might take on large solutions with many packages, given `existingPackages` is a `List<>`. Does the code still work if using a `HashSet<>`?

Reply:
Hashset seems to work fine.  Changed it.